### PR TITLE
Support Giphy embeds

### LINKED ISSUE #50: Support GIPHY
Add support for GIPHY.

https://support.giphy.com/hc/en-us/articles/360020330711-How-to-Embed-a-GIF
https://giphy.com/posts/how-to-embed-giphy-gifs-on-your-website

### PR DESCRIPTION
**Pull request by osiux**
**What**:
Support for Giphy embeds

**Why**:
Closes #50 

**How**:
Added Giphy.js and updated tests

**Checklist**:

- [x] Documentation
- [x] Tests
- [x] Ready to be merged
      

Looked on reddit and found two types of gif links, both are referenced and transformed. Didn't found a way to embed video links.

### CODE REVIEW

**Comment by MichaelDeBoey** [comment]
Hi @osiux!

Thanks for taking your time to implement the GIPHY transformer! 🙂

I'm currently on holiday until mid January, so I won't be able to review this on quickly.
Will try to look at it asap. 🙂

**Comment by osiux** [comment]
No worries, enjoy the holidays!

**Comment by MichaelDeBoey** [comment]
Hi @osiux!

This one seems a little bit more difficult when looking at different examples.

When we look at https://giphy.com/gifs/speech-martin-luther-king-jr-mlk-xTcnSXDc4tRe27yS8o/embed, we have a `margin-bottom` of `75%`.
The one you've used in your implementation uses `90%`.

Their's not really a way to know what it should be if we hard-code it, so we should look into getting it through some sort of API call I think.

**Comment by osiux** [comment]
You are right :/ even for the non responsive version it uses different heights for the iframe. Let's see if they know something and reply to your tweet :)

**Comment by MichaelDeBoey** [comment]
I hope the GIPHY team can help us out on Twitter.

https://twitter.com/MichaelDeBoey93/status/1219407708870389760

**Comment by osiux** [comment]
@MichaelDeBoey so got an idea, not sure what you think.

Basically it now uses the oembed endpoint to fetch width and height of the gif, then uses that data to calculate the padding to use to get the responsive html code. Formula is: `(height * 100 ) / width` and then round down.

Used the following data to test on their `/embed` page and [oembed](https://giphy.com/services/oembed) endpoint.

```
https://giphy.com/gifs/tom-cruise-top-gun-WDtCGLCXBAgow
w: 300
h: 151
padding: 50

https://giphy.com/gifs/jump-rabbit-mixed-LfjSCQ5ivKH3q
w: 366
h: 206
padding: 56

https://giphy.com/gifs/speech-martin-luther-king-jr-mlk-xTcnSXDc4tRe27yS8o
w: 500
h: 375
padding: 75

https://giphy.com/gifs/cute-aww-wholesome-4ZrZm6LoXmDZ7Pux3m
w: 480
h: 270
padding: 56
```

**Comment by MichaelDeBoey** [comment]
@osiux That's indeed the idea I was thinking of as a backup plan.
I was hoping there would be a way that includes less work, but unfortunately we have to work with what we have.

Let's do what you're suggesting and use the `height` & `width` we get from the oembed endpoint and  calculate the `padding-bottom` ourselves.

**Comment by osiux** [comment]
@MichaelDeBoey PR updated to do it that way :)

File: src/transformers/GIPHY.js [hunk 1]
```diff
@@ -0,0 +1,39 @@
+import { URL } from 'url';
+
+import { getTrimmedPathName, fetchOEmbedData } from './utils';
+
+const isMediaSubDomain = host => /^(media([0-9]{1,})?\.)giphy\.com$/.test(host);
```
Comment (MichaelDeBoey) [inline src/transformers/GIPHY.js]:
```suggestion
const isMediaSubDomain = host => /^(media([0-9]+)?\.)giphy\.com$/.test(host);
```

File: src/transformers/GIPHY.js [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+export const shouldTransform = url => {
+  const { host, pathname } = new URL(url);
+  const trimmedPathName = getTrimmedPathName(pathname);
+
+  return (
+    (['www.giphy.com', 'giphy.com'].includes(host) &&
+      trimmedPathName.includes('gifs/')) ||
```
Comment (MichaelDeBoey) [inline src/transformers/GIPHY.js]: No need to use `trimmedPathName`, you can just use `pathname` instead
```suggestion
      pathname.includes('/gifs/')) ||
```

File: src/transformers/GIPHY.js [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    (isMediaSubDomain(host) && trimmedPathName.includes('media/'))
```
Comment (MichaelDeBoey) [inline src/transformers/GIPHY.js]:
```suggestion
    (isMediaSubDomain(host) && pathname.includes('/media/'))
```

File: src/transformers/GIPHY.js [hunk 4]
```diff
@@ -0,0 +1,39 @@
+import { URL } from 'url';
+
+import { getTrimmedPathName, fetchOEmbedData } from './utils';
+
+const isMediaSubDomain = host => /^(media([0-9]{1,})?\.)giphy\.com$/.test(host);
+
+export const shouldTransform = url => {
+  const { host, pathname } = new URL(url);
+  const trimmedPathName = getTrimmedPathName(pathname);
```
Comment (MichaelDeBoey) [inline src/transformers/GIPHY.js]:
```suggestion
```

File: src/transformers/GIPHY.js [hunk 5]
```diff
@@ -0,0 +1,39 @@
+import { URL } from 'url';
+
+import { getTrimmedPathName, fetchOEmbedData } from './utils';
```
Comment (MichaelDeBoey) [inline src/transformers/GIPHY.js]:
```suggestion
import { fetchOEmbedData } from './utils';
```

File: src/transformers/GIPHY.js [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+  if (isMediaSubDomain(host)) {
+    return pathname.split('/')[2];
+  }
+
+  return pathname.split('-').pop();
+};
+
+export const getGIPHYResponsivePadding = ({ height, width }) =>
+  Math.floor((height / width) * 100);
```
Comment (MichaelDeBoey) [inline src/transformers/GIPHY.js]: This should be `round` instead

```suggestion
  Math.round((height / width) * 100);
```

File: src/transformers/GIPHY.js [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+
+export const getHTML = url =>
+  fetchOEmbedData(`https://giphy.com/services/oembed?url=${url}`).then(
+    ({ height, width }) => {
+      const GIPHYId = getGIPHYId(url);
+      const padding = getGIPHYResponsivePadding({ height, width });
+
+      return `<div style="width:100%;height:0;padding-bottom:${padding}%;position:relative;"><iframe src="https://giphy.com/embed/${GIPHYId}" width="100%" height="100%" style="position:absolute" frameBorder="0" class="giphy-embed" allowFullScreen></iframe></div><p><a href="https://giphy.com/gifs/${GIPHYId}">via GIPHY</a></p>`;
```
Comment (MichaelDeBoey) [inline src/transformers/GIPHY.js]: No need to add the extra `via GIPHY`.

```suggestion
      return `<div style="width:100%;height:0;padding-bottom:${padding}%;position:relative;"><iframe src="https://giphy.com/embed/${GIPHYId}" width="100%" height="100%" style="position:absolute" frameborder="0" class="giphy-embed" allowfullscreen></iframe></div>`;
```

**Comment by osiux** [comment]
@MichaelDeBoey Duh, missed those two haha should be good now.

**Comment by MichaelDeBoey** [comment]
@all-contributors Please add @osiux for code and tests

**Pull request merged**